### ISSUE TITLE
Nullable doesn't warn for object creations involving dynamic in nested initializers

### ISSUE THREAD

**Issue by 333fred** [dotnet/roslyn#45088]
```cs
#nullable enable
class A
{
    dynamic this[int x]
    {
        get
        {
            return new A();
        }
    }
    static void Main()
    {
        object? o = null;
        dynamic x = 1;
        A _ = new A() { [x] = { AnyProp = o.ToString() } };
    }
}
```

We should warn on the `o.ToString()`, and indeed we do if a different compilation error is introduced that causes the initializer to be bound as a bad expression instead. The culprit is here: https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs#L2657.
